From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Max Filippov <jcmvbkbc@gmail.com>
Date: Tue, 10 Oct 2023 17:02:39 -0700
Subject: [PATCH 22/61] drivers/net/wireless/espressif/esp32-ng: update

Update to upstream commit 6a25417fc880fd744b3b0d93c11659c3e7d86384

Signed-off-by: Max Filippov <jcmvbkbc@gmail.com>
(cherry picked from commit abeb40c333c149edac3aa4cb85e05b96cd426000)
---
 .../net/wireless/espressif/esp32-ng/Makefile  |   2 +-
 .../net/wireless/espressif/esp32-ng/adapter.h |   6 +-
 drivers/net/wireless/espressif/esp32-ng/esp.h |   5 +
 .../net/wireless/espressif/esp32-ng/esp_api.h |  10 +-
 .../net/wireless/espressif/esp32-ng/esp_bt.c  |   2 +
 .../espressif/esp32-ng/esp_cfg80211.c         | 105 +++++---
 .../espressif/esp32-ng/esp_cfg80211.h         |   4 +-
 .../net/wireless/espressif/esp32-ng/esp_cmd.c |  98 ++++++--
 .../net/wireless/espressif/esp32-ng/esp_cmd.h |   1 +
 .../net/wireless/espressif/esp32-ng/esp_if.h  |   2 +-
 .../espressif/esp32-ng/esp_kernel_port.h      |   5 +
 .../wireless/espressif/esp32-ng/esp_sdio.c    | 158 ++++++------
 .../espressif/esp32-ng/esp_sdio_decl.h        |  14 +-
 .../net/wireless/espressif/esp32-ng/esp_spi.c | 191 +++++----------
 .../wireless/espressif/esp32-ng/esp_stats.c   |  21 +-
 .../wireless/espressif/esp32-ng/esp_stats.h   |   4 -
 .../esp32-ng/{esp_wpa_utils.c => esp_utils.c} |  43 ++--
 .../esp32-ng/{esp_wpa_utils.h => esp_utils.h} |   5 +-
 .../net/wireless/espressif/esp32-ng/main.c    | 224 +++++++++++++-----
 .../net/wireless/espressif/esp32-ng/utils.h   |   4 +
 20 files changed, 523 insertions(+), 381 deletions(-)
 rename drivers/net/wireless/espressif/esp32-ng/{esp_wpa_utils.c => esp_utils.c} (55%)
 rename drivers/net/wireless/espressif/esp32-ng/{esp_wpa_utils.h => esp_utils.h} (92%)

diff --git a/drivers/net/wireless/espressif/esp32-ng/Makefile b/drivers/net/wireless/espressif/esp32-ng/Makefile
index 22df66eaa1bb5..14667bf6d5916 100644
--- a/drivers/net/wireless/espressif/esp32-ng/Makefile
+++ b/drivers/net/wireless/espressif/esp32-ng/Makefile
@@ -1,5 +1,5 @@
 obj-$(CONFIG_ESP32_WIFI)		+= esp32-wifi.o
-esp32-wifi-objs				+= main.o esp_bt.o esp_cmd.o esp_wpa_utils.o esp_cfg80211.o esp_stats.o
+esp32-wifi-objs				+= main.o esp_bt.o esp_cmd.o esp_cfg80211.o esp_stats.o esp_utils.o
 esp32-wifi-$(CONFIG_ESP32_WIFI_SPI)	+= esp_spi.o
 esp32-wifi-$(CONFIG_ESP32_WIFI_SDIO)	+= esp_sdio.o
 esp32-wifi-objs				+= $(esp32-wifi-m)
diff --git a/drivers/net/wireless/espressif/esp32-ng/adapter.h b/drivers/net/wireless/espressif/esp32-ng/adapter.h
index 0efe5ddac3595..783d08bd71f97 100644
--- a/drivers/net/wireless/espressif/esp32-ng/adapter.h
+++ b/drivers/net/wireless/espressif/esp32-ng/adapter.h
@@ -85,8 +85,8 @@ enum ESP_CAPABILITIES {
 };
 
 typedef enum {
-	ESP_TEST_RAW_TP = (1 << 0),
-	ESP_TEST_RAW_TP__ESP_TO_HOST = (1 << 1)
+	ESP_TEST_RAW_TP_HOST_TO_ESP = (1 << 0),
+	ESP_TEST_RAW_TP_ESP_TO_HOST = (1 << 1)
 } ESP_RAW_TP_MEASUREMENT;
 
 enum ESP_INTERNAL_MSG {
@@ -120,6 +120,8 @@ enum COMMAND_CODE {
 	CMD_SET_TXPOWER,
 	CMD_GET_REG_DOMAIN,
 	CMD_SET_REG_DOMAIN,
+	CMD_RAW_TP_ESP_TO_HOST,
+	CMD_RAW_TP_HOST_TO_ESP,
 	CMD_MAX,
 };
 
diff --git a/drivers/net/wireless/espressif/esp32-ng/esp.h b/drivers/net/wireless/espressif/esp32-ng/esp.h
index c391ac9804679..c320214bd765b 100644
--- a/drivers/net/wireless/espressif/esp32-ng/esp.h
+++ b/drivers/net/wireless/espressif/esp32-ng/esp.h
@@ -40,6 +40,8 @@ enum chipset_type_e {
 	ESP_FIRMWARE_CHIP_ESP32S2 = 0x2,
 	ESP_FIRMWARE_CHIP_ESP32C3 = 0x5,
 	ESP_FIRMWARE_CHIP_ESP32S3 = 0x9,
+	ESP_FIRMWARE_CHIP_ESP32C2 = 0x0C,
+	ESP_FIRMWARE_CHIP_ESP32C6 = 0x0D,
 };
 
 #define ESP_PAYLOAD_HEADER      8
@@ -58,6 +60,7 @@ enum adapter_flags_e {
 	ESP_CLEANUP_IN_PROGRESS,    /* Driver unloading or ESP reseted */
 	ESP_CMD_INIT_DONE,          /* Cmd component is initialized with esp_commands_setup() */
 	ESP_DRIVER_ACTIVE,          /* kernel module __exit is not yet invoked */
+	ESP_INIT_DONE,              /* Driver init done */
 };
 
 enum priv_flags_e {
@@ -115,6 +118,7 @@ struct esp_adapter {
 	struct work_struct      events_work;
 
 	unsigned long           state_flags;
+	int                     chipset;
 };
 
 struct esp_device {
@@ -158,6 +162,7 @@ struct esp_wifi_device {
 	struct notifier_block   nb;
 	uint8_t                 tx_pwr_type;
 	uint8_t                 tx_pwr;
+	uint32_t                rssi;
 };
 
 
diff --git a/drivers/net/wireless/espressif/esp32-ng/esp_api.h b/drivers/net/wireless/espressif/esp32-ng/esp_api.h
index 6543212354dd8..737524b5e584d 100644
--- a/drivers/net/wireless/espressif/esp32-ng/esp_api.h
+++ b/drivers/net/wireless/espressif/esp32-ng/esp_api.h
@@ -14,7 +14,7 @@
 	do {                                                          \
 		int ret = (x);                                        \
 		if (!!(ret)) {                                        \
-		    esp_err("esp32: %s failure, ret: %d\n", #x, ret); \
+		    esp_err("%s failure, ret: %d\n", #x, ret);        \
 		    return ret;                                       \
 		}                                                     \
 	} while (0);                                                  \
@@ -29,14 +29,16 @@ int esp_send_packet(struct esp_adapter *adapter, struct sk_buff *skb);
 u8 esp_is_bt_supported_over_sdio(u32 cap);
 void esp_tx_pause(struct esp_wifi_device *priv);
 void esp_tx_resume(struct esp_wifi_device *priv);
-void process_event_esp_bootup(struct esp_adapter *adapter, u8 *evt_buf, u8 len);
-int process_fw_data(struct fw_data *fw_p);
 void esp_init_priv(struct net_device *ndev);
 void esp_port_open(struct esp_wifi_device *priv);
 void esp_port_close(struct esp_wifi_device *priv);
 void esp_remove_network_interfaces(struct esp_adapter *adapter);
 void print_capabilities(u32 cap);
 void process_capabilities(struct esp_adapter *adapter);
-void process_test_capabilities(u8 cap);
 int esp_is_tx_queue_paused(struct esp_wifi_device *priv);
+int esp_deinit_module(struct esp_adapter *adapter);
+int esp_validate_chipset(struct esp_adapter *adapter, u8 chipset);
+int esp_adjust_spi_clock(struct esp_adapter *adapter, u8 spi_clk_mhz);
+void process_test_capabilities(u32 raw_tp_mode);
+int esp_init_raw_tp(struct esp_adapter *adapter);
 #endif
diff --git a/drivers/net/wireless/espressif/esp32-ng/esp_bt.c b/drivers/net/wireless/espressif/esp32-ng/esp_bt.c
index 9fdc08af23f27..7074ed56453fc 100644
--- a/drivers/net/wireless/espressif/esp32-ng/esp_bt.c
+++ b/drivers/net/wireless/espressif/esp32-ng/esp_bt.c
@@ -242,6 +242,8 @@ int esp_init_bt(struct esp_adapter *adapter)
 
 	hdev->dev_type = HCI_PRIMARY;
 
+	SET_HCIDEV_DEV(hdev, adapter->dev);
+
 	ret = hci_register_dev(hdev);
 	if (ret < 0) {
 		BT_ERR("Can not register HCI device");
diff --git a/drivers/net/wireless/espressif/esp32-ng/esp_cfg80211.c b/drivers/net/wireless/espressif/esp32-ng/esp_cfg80211.c
index 5273d560995f0..feeb9921a198d 100644
--- a/drivers/net/wireless/espressif/esp32-ng/esp_cfg80211.c
+++ b/drivers/net/wireless/espressif/esp32-ng/esp_cfg80211.c
@@ -260,7 +260,6 @@ static int esp_cfg80211_scan(struct wiphy *wiphy,
 
 	struct net_device *ndev = NULL;
 	struct esp_wifi_device *priv = NULL;
-
 	if (!wiphy || !request || !request->wdev || !request->wdev->netdev) {
 		esp_info("%u invalid input\n", __LINE__);
 		return -EINVAL;
@@ -323,21 +322,7 @@ static int esp_cfg80211_del_key(struct wiphy *wiphy, struct net_device *dev,
 				INT_LINK_ID u8 key_index, bool pairwise,
 				const u8 *mac_addr)
 {
-	struct esp_wifi_device *priv = NULL;
-
-	if (!wiphy || !dev) {
-		esp_err("%u invalid params\n", __LINE__);
-		return -EINVAL;
-	}
-
-	priv = netdev_priv(dev);
-	if (!priv) {
-		esp_err("Empty priv\n");
-		return -EINVAL;
-	}
-	esp_dbg("\n");
-
-	return cmd_del_key(priv, key_index, pairwise, mac_addr);
+	return 0;
 }
 
 static int esp_cfg80211_add_key(struct wiphy *wiphy, struct net_device *dev,
@@ -554,6 +539,39 @@ static int esp_cfg80211_set_tx_power(struct wiphy *wiphy,
 	return cmd_set_tx_power(priv, priv->tx_pwr);
 }
 
+static int esp_cfg80211_get_station(struct wiphy *wiphy, struct net_device *ndev,
+				    const u8 *mac, struct station_info *sinfo)
+{
+	struct esp_wifi_device *priv = NULL;
+
+	priv = netdev_priv(ndev);
+
+	if (!mac || !priv) {
+		esp_err("mac=%p priv=%p\n", mac, priv);
+		return -ENOENT;
+	}
+	if (wireless_dev_current_bss_exists(&priv->wdev)) {
+
+		sinfo->filled |= BIT(NL80211_STA_INFO_SIGNAL);
+		sinfo->signal = priv->rssi;
+
+		sinfo->filled |= BIT(NL80211_STA_INFO_RX_BYTES);
+		sinfo->rx_bytes = priv->stats.rx_bytes;
+		sinfo->filled |= BIT(NL80211_STA_INFO_RX_PACKETS);
+		sinfo->rx_packets = priv->stats.rx_packets;
+
+		sinfo->filled |= BIT(NL80211_STA_INFO_TX_BYTES);
+		sinfo->tx_bytes = priv->stats.tx_bytes;
+		sinfo->filled |= BIT(NL80211_STA_INFO_TX_PACKETS);
+		sinfo->tx_packets = priv->stats.tx_packets;
+
+		sinfo->filled |= BIT(NL80211_STA_INFO_TX_FAILED);
+		sinfo->tx_failed = priv->stats.tx_dropped;
+	}
+
+	return 0;
+}
+
 static int esp_cfg80211_get_tx_power(struct wiphy *wiphy,
 				     struct wireless_dev *wdev,
 				     int *dbm)
@@ -600,6 +618,7 @@ static struct cfg80211_ops esp_cfg80211_ops = {
 	.set_wakeup = esp_cfg80211_set_wakeup,
 	.set_tx_power = esp_cfg80211_set_tx_power,
 	.get_tx_power = esp_cfg80211_get_tx_power,
+	.get_station = esp_cfg80211_get_station,
 };
 
 static void esp_reg_notifier(struct wiphy *wiphy,
@@ -614,15 +633,21 @@ static void esp_reg_notifier(struct wiphy *wiphy,
 		return;
 	}
 
+	if (!test_bit(ESP_INIT_DONE, &adapter->state_flags)) {
+		esp_info("Driver init is ongoing\n");
+		return;
+	}
+	if (test_bit(ESP_CLEANUP_IN_PROGRESS, &adapter->state_flags)) {
+		esp_info("Driver cleanup is ongoing\n");
+		return;
+	}
+
 	esp_dev = wiphy_priv(wiphy);
 
 	if (!esp_dev || !esp_dev->adapter) {
 		esp_info("%u esp_dev not initialized yet \n", __LINE__);
 		return;
 	}
-	if (test_bit(ESP_CLEANUP_IN_PROGRESS, &adapter->state_flags)) {
-               return;
-	}
 
 	priv = esp_dev->adapter->priv[0];
 
@@ -646,17 +671,22 @@ static void esp_reg_notifier(struct wiphy *wiphy,
 
 	/* Don't send same regdom info to firmware */
 	if (strncmp(request->alpha2, priv->country_code, strlen(request->alpha2))) {
-		strlcpy(priv->country_code, request->alpha2, MAX_COUNTRY_LEN);
+		strscpy(priv->country_code, request->alpha2, MAX_COUNTRY_LEN);
 		cmd_set_reg_domain(priv);
 	}
 }
 
-int esp_cfg80211_register(struct esp_adapter *adapter)
+int esp_add_wiphy(struct esp_adapter *adapter)
 {
 	struct wiphy *wiphy;
 	struct esp_device *esp_dev;
 	int ret = 0;
 
+	if (!adapter) {
+		esp_info("adapter not yet initialized\n");
+		return -EINVAL;
+	}
+
 	wiphy = wiphy_new(&esp_cfg80211_ops, sizeof(struct esp_device));
 
 	if (!wiphy) {
@@ -701,28 +731,35 @@ int esp_cfg80211_register(struct esp_adapter *adapter)
 	return ret;
 }
 
-int esp_mark_disconnect(struct esp_wifi_device *priv, uint16_t reason,
-		uint8_t locally_disconnect)
+int esp_remove_wiphy(struct esp_adapter *adapter)
 {
-	if (priv && priv->ndev)
-		if (priv->ndev->reg_state == NETREG_REGISTERED)
-			CFG80211_DISCONNECTED(priv->ndev, reason, NULL, 0, locally_disconnect,
-					GFP_KERNEL);
+	if (adapter && adapter->wiphy) {
+		wiphy_unregister(adapter->wiphy);
+		wiphy_free(adapter->wiphy);
+		adapter->wiphy = NULL;
+	}
+
+	return 0;
+}
+
+int esp_mark_disconnect(struct esp_wifi_device *priv, uint16_t reason, uint8_t locally_disconnect)
+{
+	if (priv && priv->ndev && wireless_dev_current_bss_exists(&priv->wdev))
+		CFG80211_DISCONNECTED(priv->ndev, reason, NULL, 0, locally_disconnect, GFP_KERNEL);
+
 	return 0;
 }
 
-int esp_mark_scan_done_and_disconnect(struct esp_wifi_device *priv,
-		uint8_t locally_disconnect)
+int esp_mark_scan_done_and_disconnect(struct esp_wifi_device *priv, uint8_t locally_disconnect)
 {
 
 	if (!priv)
 		return -EINVAL;
 
-	ESP_MARK_SCAN_DONE(priv, true);
+	if (priv->wdev.iftype != NL80211_IFTYPE_STATION)
+		return 0;
 
+	ESP_MARK_SCAN_DONE(priv, true);
 	ESP_CANCEL_SCHED_SCAN();
-
-	esp_mark_disconnect(priv, 0, locally_disconnect);
-
-	return 0;
+	return esp_mark_disconnect(priv, 0, locally_disconnect);
 }
diff --git a/drivers/net/wireless/espressif/esp32-ng/esp_cfg80211.h b/drivers/net/wireless/espressif/esp32-ng/esp_cfg80211.h
index a669011c2ac94..05f71f2598269 100644
--- a/drivers/net/wireless/espressif/esp32-ng/esp_cfg80211.h
+++ b/drivers/net/wireless/espressif/esp32-ng/esp_cfg80211.h
@@ -15,8 +15,8 @@ struct wireless_dev *esp_cfg80211_add_iface(struct wiphy *wiphy,
 			      unsigned char name_assign_type,
 			      enum nl80211_iftype type,
 			      struct vif_params *params);
-int esp_cfg80211_register(struct esp_adapter *adapter);
-
+int esp_add_wiphy(struct esp_adapter *adapter);
+int esp_remove_wiphy(struct esp_adapter *adapter);
 int esp_mark_disconnect(struct esp_wifi_device *priv, uint16_t reason,
 		uint8_t locally_disconnect);
 int esp_mark_scan_done_and_disconnect(struct esp_wifi_device *priv,
diff --git a/drivers/net/wireless/espressif/esp32-ng/esp_cmd.c b/drivers/net/wireless/espressif/esp32-ng/esp_cmd.c
index 29bc3c99e64d4..0c0ab968ef66b 100644
--- a/drivers/net/wireless/espressif/esp32-ng/esp_cmd.c
+++ b/drivers/net/wireless/espressif/esp32-ng/esp_cmd.c
@@ -8,16 +8,18 @@
 #include "utils.h"
 #include "esp_cmd.h"
 #include "esp_api.h"
-#include "esp_wpa_utils.h"
+#include "esp_utils.h"
 #include "esp.h"
 #include "esp_cfg80211.h"
 #include "esp_kernel_port.h"
+#include "esp_stats.h"
 
 #define PRINT_HEXDUMP(STR, ARG, ARG_LEN, level) \
 	print_hex_dump(KERN_INFO, STR, DUMP_PREFIX_ADDRESS, 16, 1, ARG, ARG_LEN, 1);
 
 #define COMMAND_RESPONSE_TIMEOUT (5 * HZ)
 u8 ap_bssid[MAC_ADDR_LEN];
+extern u32 raw_tp_mode;
 
 int internal_scan_request(struct esp_wifi_device *priv, char *ssid,
 		uint8_t channel, uint8_t is_blocking);
@@ -235,6 +237,8 @@ static int wait_and_decode_cmd_resp(struct esp_wifi_device *priv,
 	case CMD_SET_MCAST_MAC_ADDR:
 	case CMD_GET_REG_DOMAIN:
 	case CMD_SET_REG_DOMAIN:
+	case CMD_RAW_TP_ESP_TO_HOST:
+	case CMD_RAW_TP_HOST_TO_ESP:
 		/* intentional fallthrough */
 		if (ret == 0)
 			ret = decode_common_resp(cmd_node);
@@ -426,6 +430,10 @@ struct command_node *prepare_command_request(struct esp_adapter *adapter, u8 cmd
 		esp_err("unsupported command code\n");
 		return NULL;
 	}
+	if (!test_bit(ESP_CMD_INIT_DONE, &adapter->state_flags)) {
+		esp_err("command queue init is not done yet\n");
+		return NULL;
+	}
 
 	node = get_free_cmd_node(adapter);
 
@@ -450,7 +458,6 @@ struct command_node *prepare_command_request(struct esp_adapter *adapter, u8 cmd
 	cmd->cmd_code = cmd_code;
 
 /*	payload_header->checksum = cpu_to_le16(compute_checksum(skb->data, len));*/
-
 	return node;
 }
 
@@ -568,6 +575,23 @@ static void process_auth_event(struct esp_wifi_device *priv,
 
 }
 
+#define IEEE80211_DEAUTH_FRAME_LEN      (24 /* hdr */ + 2 /* reason */)
+static void process_deauth_event(struct esp_wifi_device *priv, struct disconnect_event *event)
+{
+	u8 frame_buf[IEEE80211_DEAUTH_FRAME_LEN];
+	struct ieee80211_mgmt *mgmt = (void *)frame_buf;
+
+	/* build frame */
+	mgmt->frame_control = cpu_to_le16(IEEE80211_FTYPE_MGMT | IEEE80211_STYPE_DEAUTH);
+	mgmt->duration = 0; /* initialize only */
+	mgmt->seq_ctrl = 0; /* initialize only */
+	memcpy(mgmt->da, priv->mac_address, ETH_ALEN); /* own address */
+	memcpy(mgmt->sa, event->bssid, ETH_ALEN);
+	memcpy(mgmt->bssid, event->bssid, ETH_ALEN);
+	mgmt->u.deauth.reason_code = cpu_to_le16(event->reason);
+	cfg80211_rx_mlme_mgmt(priv->ndev, frame_buf, IEEE80211_DEAUTH_FRAME_LEN);
+}
+
 static void process_disconnect_event(struct esp_wifi_device *priv,
 		struct disconnect_event *event)
 {
@@ -580,6 +604,12 @@ static void process_disconnect_event(struct esp_wifi_device *priv,
 			event->ssid, event->reason);
 
 	esp_mark_disconnect(priv, event->reason, true);
+	/* Flush previous scan results from kernel */
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(5, 9, 0))
+	cfg80211_bss_flush(priv->adapter->wiphy);
+#endif
+	/* Send dummpy deauth to userspace */
+	process_deauth_event(priv, event);
 }
 
 static void process_assoc_event(struct esp_wifi_device *priv,
@@ -595,6 +625,7 @@ static void process_assoc_event(struct esp_wifi_device *priv,
 	esp_info("Connection status: %d\n", event->header.status);
 
 	memcpy(mac, event->bssid, MAC_ADDR_LEN);
+	priv->rssi = event->rssi;
 
 	CFG80211_RX_ASSOC_RESP(priv->ndev, priv->bss, event->frame, event->frame_len,
 			0, priv->assoc_req_ie, priv->assoc_req_ie_len);
@@ -1171,6 +1202,15 @@ int cmd_add_key(struct esp_wifi_device *priv, u8 key_index, bool pairwise,
 	}
 #endif
 
+       /* Supplicant swaps tx/rx Mic keys whereas esp needs it normal format */
+       if (key->algo == WIFI_WPA_ALG_TKIP && !key->index) {
+               u8 buf[8];
+               memcpy(buf, &key->data[16], 8);
+               memcpy(&key->data[16], &key->data[24], 8);
+               memcpy(&key->data[24], buf, 8);
+               memset(buf, 0, 8);
+       }
+
 #if 0
 	esp_err("%u algo: %u idx: %u seq_len: %u len:%u\n", __LINE__,
 			key->algo, key->index, key->seq_len, key->len);
@@ -1368,6 +1408,36 @@ int cmd_scan_request(struct esp_wifi_device *priv, struct cfg80211_scan_request
 	return 0;
 }
 
+int cmd_init_raw_tp_task_timer(struct esp_wifi_device *priv)
+{
+	u16 cmd_len;
+	struct command_node *cmd_node = NULL;
+
+	if (!priv || !priv->adapter) {
+		esp_err("Invalid argument\n");
+		return -EINVAL;
+	}
+
+	cmd_len = sizeof(struct command_header);
+
+	if (raw_tp_mode == ESP_TEST_RAW_TP_ESP_TO_HOST) {
+		cmd_node = prepare_command_request(priv->adapter, CMD_RAW_TP_ESP_TO_HOST, cmd_len);
+	} else if (raw_tp_mode == ESP_TEST_RAW_TP_HOST_TO_ESP) {
+		cmd_node = prepare_command_request(priv->adapter, CMD_RAW_TP_HOST_TO_ESP, cmd_len);
+	}
+
+	if (!cmd_node) {
+		esp_err("Failed to get command node\n");
+		return -ENOMEM;
+	}
+
+	queue_cmd_node(priv->adapter, cmd_node, ESP_CMD_DFLT_PRIO);
+	queue_work(priv->adapter->cmd_wq, &priv->adapter->cmd_work);
+
+	RET_ON_FAIL(wait_and_decode_cmd_resp(priv, cmd_node));
+
+	return 0;
+}
 
 int cmd_get_mac(struct esp_wifi_device *priv)
 {
@@ -1430,32 +1500,12 @@ int cmd_set_mac(struct esp_wifi_device *priv, uint8_t *mac_addr)
 
 int esp_commands_teardown(struct esp_adapter *adapter)
 {
-#define MAX_DEINIT_RETRY 5
-	uint8_t iface_idx = 0;
-	struct esp_wifi_device *priv = NULL;
-
 	if (!adapter) {
 		return -EINVAL;
 	}
 
 	set_bit(ESP_CLEANUP_IN_PROGRESS, &adapter->state_flags);
-
-	if (!test_bit(ESP_CMD_INIT_DONE, &adapter->state_flags))
-		return 0;
-
-
-	for (iface_idx = 0; iface_idx < ESP_MAX_INTERFACE; iface_idx++) {
-
-		priv = adapter->priv[iface_idx];
-
-		if (!priv)
-			continue;
-
-		esp_mark_scan_done_and_disconnect(priv, false);
-
-		esp_port_close(priv);
-	}
-
+	clear_bit(ESP_CMD_INIT_DONE, &adapter->state_flags);
 	destroy_cmd_wq(adapter);
 	free_esp_cmd_pool(adapter);
 
@@ -1596,7 +1646,7 @@ int cmd_set_reg_domain(struct esp_wifi_device *priv)
 
 	cmd = (struct cmd_reg_domain *) (cmd_node->cmd_skb->data + sizeof(struct esp_payload_header));
 
-	strlcpy(cmd->country_code, priv->country_code, MAX_COUNTRY_LEN);
+	strscpy(cmd->country_code, priv->country_code, MAX_COUNTRY_LEN);
 
 	queue_cmd_node(priv->adapter, cmd_node, ESP_CMD_DFLT_PRIO);
 	queue_work(priv->adapter->cmd_wq, &priv->adapter->cmd_work);
diff --git a/drivers/net/wireless/espressif/esp32-ng/esp_cmd.h b/drivers/net/wireless/espressif/esp32-ng/esp_cmd.h
index 6e9e0aa156223..393c13b300232 100644
--- a/drivers/net/wireless/espressif/esp32-ng/esp_cmd.h
+++ b/drivers/net/wireless/espressif/esp32-ng/esp_cmd.h
@@ -51,4 +51,5 @@ int cmd_set_tx_power(struct esp_wifi_device *priv, int power);
 int cmd_get_tx_power(struct esp_wifi_device *priv);
 int cmd_set_reg_domain(struct esp_wifi_device *priv);
 int cmd_get_reg_domain(struct esp_wifi_device *priv);
+int cmd_init_raw_tp_task_timer(struct esp_wifi_device *priv);
 #endif
diff --git a/drivers/net/wireless/espressif/esp32-ng/esp_if.h b/drivers/net/wireless/espressif/esp32-ng/esp_if.h
index 47ff7136d78d3..78cb457b146a6 100644
--- a/drivers/net/wireless/espressif/esp32-ng/esp_if.h
+++ b/drivers/net/wireless/espressif/esp32-ng/esp_if.h
@@ -17,7 +17,7 @@ struct esp_if_ops {
 	int (*deinit)(struct esp_adapter *adapter);
 };
 
-int esp_init_interface_layer(struct esp_adapter *adapter);
+int esp_init_interface_layer(struct esp_adapter *adapter, u32 speed);
 void esp_deinit_interface_layer(void);
 
 #endif
diff --git a/drivers/net/wireless/espressif/esp32-ng/esp_kernel_port.h b/drivers/net/wireless/espressif/esp32-ng/esp_kernel_port.h
index 03d11773e54ba..c4eb336969911 100644
--- a/drivers/net/wireless/espressif/esp32-ng/esp_kernel_port.h
+++ b/drivers/net/wireless/espressif/esp32-ng/esp_kernel_port.h
@@ -178,7 +178,12 @@ void CFG80211_RX_ASSOC_RESP(struct net_device *dev,
 			    const u8 *req_ies, size_t req_ies_len)
 {
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 0, 0))
+
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 7, 0))
+	struct cfg80211_rx_assoc_resp_data resp = {0};
+#else
 	struct cfg80211_rx_assoc_resp resp = {0};
+#endif
 
 	resp.links[0].bss = bss;
 	resp.buf = (u8 *)buf;
diff --git a/drivers/net/wireless/espressif/esp32-ng/esp_sdio.c b/drivers/net/wireless/espressif/esp32-ng/esp_sdio.c
index c986ecf7927f1..ee0cf5af95b88 100644
--- a/drivers/net/wireless/espressif/esp32-ng/esp_sdio.c
+++ b/drivers/net/wireless/espressif/esp32-ng/esp_sdio.c
@@ -18,8 +18,10 @@
 #include "esp_bt_api.h"
 #include <linux/kthread.h>
 #include "esp_stats.h"
+#include "esp_utils.h"
 #include "include/esp_kernel_port.h"
 
+extern u32 raw_tp_mode;
 #define MAX_WRITE_RETRIES       2
 #define TX_MAX_PENDING_COUNT    200
 #define TX_RESUME_THRESHOLD     (TX_MAX_PENDING_COUNT/5)
@@ -164,6 +166,16 @@ static int esp_slave_get_tx_buffer_num(struct esp_sdio_context *context, u32 *tx
 	return ret;
 }
 
+int esp_deinit_module(struct esp_adapter *adapter)
+{
+	/* Second & onward bootup cleanup is not required for SDIO:
+	 * As Removal of SDIO triggers complete Deinit and SDIO insertion/
+	 * detection, triggers probing which does initialization.
+	 */
+
+	return 0;
+}
+
 static int esp_get_len_from_slave(struct esp_sdio_context *context, u32 *rx_size, u8 is_lock_needed)
 {
 	u32 *len;
@@ -269,6 +281,7 @@ static void esp_remove(struct sdio_func *func)
 		}
 		memset(context, 0, sizeof(struct esp_sdio_context));
 	}
+	esp_dbg("ESP SDIO cleanup completed\n");
 }
 
 static struct esp_if_ops if_ops = {
@@ -598,9 +611,11 @@ static int tx_process(void *data)
 		cb = (struct esp_skb_cb *)tx_skb->cb;
 		if (cb && cb->priv && atomic_read(&tx_pending) < TX_RESUME_THRESHOLD) {
 			esp_tx_resume(cb->priv);
-			#if TEST_RAW_TP
+#if TEST_RAW_TP
+			if (raw_tp_mode != 0) {
 				esp_raw_tp_queue_resume();
-			#endif
+			}
+#endif
 		}
 
 		buf_needed = (tx_skb->len + ESP_RX_BUFFER_SIZE - 1) / ESP_RX_BUFFER_SIZE;
@@ -653,7 +668,7 @@ static int tx_process(void *data)
 	return 0;
 }
 
-static struct esp_sdio_context *init_sdio_func(struct sdio_func *func)
+static struct esp_sdio_context *init_sdio_func(struct sdio_func *func, int *sdio_ret)
 {
 	struct esp_sdio_context *context = NULL;
 	int ret = 0;
@@ -670,13 +685,24 @@ static struct esp_sdio_context *init_sdio_func(struct sdio_func *func)
 	/* Enable Function */
 	ret = sdio_enable_func(func);
 	if (ret) {
+		esp_err("sdio_enable_func ret: %d\n", ret);
+		if (sdio_ret)
+			*sdio_ret = ret;
+		sdio_release_host(func);
+
 		return NULL;
 	}
 
 	/* Register IRQ */
 	ret = sdio_claim_irq(func, esp_handle_isr);
 	if (ret) {
+		esp_err("sdio_claim_irq ret: %d\n", ret);
 		sdio_disable_func(func);
+
+		if (sdio_ret)
+			*sdio_ret = ret;
+		sdio_release_host(func);
+
 		return NULL;
 	}
 
@@ -757,10 +783,25 @@ static int esp_probe(struct sdio_func *func,
 
 	esp_info("ESP network device detected\n");
 
-	context = init_sdio_func(func);
+	context = init_sdio_func(func, &ret);;
 
 	if (!context) {
-		return -ENOMEM;
+		if (ret)
+			return ret;
+		else
+			return -EINVAL;
+	}
+
+	if (sdio_context.sdio_clk_mhz) {
+		struct mmc_host *host = func->card->host;
+		u32 hz = sdio_context.sdio_clk_mhz * NUMBER_1M;
+		/* Expansion of mmc_set_clock that isnt exported */
+		if (hz < host->f_min)
+			hz = host->f_min;
+		if (hz > host->f_max)
+			hz = host->f_max;
+		host->ios.clock = hz;
+		host->ops->set_ios(host, &host->ios);
 	}
 
 	context->state = ESP_CONTEXT_READY;
@@ -771,10 +812,10 @@ static int esp_probe(struct sdio_func *func,
 		return ret;
 	}
 
-	tx_thread = kthread_run(tx_process, context->adapter, "esp32_TX");
+	tx_thread = kthread_run(tx_process, context->adapter, "esp_TX");
 
 	if (!tx_thread)
-		esp_err("Failed to create esp32_sdio TX thread\n");
+		esp_err("Failed to create esp_sdio TX thread\n");
 
 	context->adapter->dev = &func->dev;
 	generate_slave_intr(context, BIT(ESP_OPEN_DATA_PATH));
@@ -787,6 +828,8 @@ static int esp_probe(struct sdio_func *func,
 		esp_err("Failed to create monitor thread\n");
 #endif
 
+	esp_dbg("ESP SDIO probe completed\n");
+
 	return ret;
 }
 
@@ -866,20 +909,27 @@ static const struct dev_pm_ops esp_pm_ops = {
 	.resume = esp_resume,
 };
 
+static const struct of_device_id esp_sdio_of_match[] = {
+	{ .compatible = "espressif,esp_sdio", },
+	{ /* sentinel */ }
+};
+MODULE_DEVICE_TABLE(of, esp_sdio_of_match);
+
 /* SDIO driver structure to be registered with kernel */
 static struct sdio_driver esp_sdio_driver = {
-	.name		= "esp_sdio",
+	.name		= KBUILD_MODNAME,
 	.id_table	= esp_devices,
 	.probe		= esp_probe,
 	.remove		= esp_remove,
 	.drv = {
+		.name = KBUILD_MODNAME,
 		.owner = THIS_MODULE,
 		.pm = &esp_pm_ops,
-	}
-
+		.of_match_table = esp_sdio_of_match,
+	},
 };
 
-int esp_init_interface_layer(struct esp_adapter *adapter)
+int esp_init_interface_layer(struct esp_adapter *adapter, u32 speed)
 {
 	if (!adapter)
 		return -EINVAL;
@@ -887,73 +937,41 @@ int esp_init_interface_layer(struct esp_adapter *adapter)
 	adapter->if_context = &sdio_context;
 	adapter->if_ops = &if_ops;
 	sdio_context.adapter = adapter;
+	sdio_context.sdio_clk_mhz = speed;
 
 	return sdio_register_driver(&esp_sdio_driver);
 }
 
-void process_event_esp_bootup(struct esp_adapter *adapter, u8 *evt_buf, u8 len)
+int esp_validate_chipset(struct esp_adapter *adapter, u8 chipset)
 {
-	u8 len_left = len, tag_len;
-	u8 *pos;
-	struct esp_sdio_context *context = &sdio_context;
-
-	if (!adapter)
-		return;
-
-	if (!evt_buf)
-		return;
-
-	pos = evt_buf;
-
-	while (len_left) {
-		tag_len = *(pos + 1);
-
-		esp_info("EVENT: %d\n", *pos);
-
-		if (*pos == ESP_BOOTUP_CAPABILITY) {
-
-			adapter->capabilities = *(pos + 2);
-			process_capabilities(adapter);
-			print_capabilities(*(pos + 2));
-
-		} else if (*pos == ESP_BOOTUP_FIRMWARE_CHIP_ID) {
-
-			esp_info("ESP chipset detected [%s]\n",
-				*(pos+2) == ESP_FIRMWARE_CHIP_ESP32 ? "esp32" :
-				*(pos+2) == ESP_FIRMWARE_CHIP_ESP32S2 ? "esp32-s2" :
-				*(pos+2) == ESP_FIRMWARE_CHIP_ESP32C3 ? "esp32-c3" :
-				*(pos+2) == ESP_FIRMWARE_CHIP_ESP32S3 ? "esp32-s3" :
-				"unknown");
-
-			if (*(pos+2) != ESP_FIRMWARE_CHIP_ESP32)
-				esp_err("SDIO is only supported with ESP32\n");
-
-		} else if (*pos == ESP_BOOTUP_TEST_RAW_TP) {
-			process_test_capabilities(*(pos + 2));
-
-		} else if (*pos == ESP_BOOTUP_FW_DATA) {
-
-			if (tag_len != sizeof(struct fw_data))
-				esp_info("Length not matching to firmware data size\n");
-			else
-				if (process_fw_data((struct fw_data *)(pos + 2)))
-					if (context->func) {
-						generate_slave_intr(context, BIT(ESP_CLOSE_DATA_PATH));
-						return;
-					}
-
-		} else {
-			esp_warn("Unsupported tag in event");
-		}
+	int ret = 0;
 
-		pos += (tag_len+2);
-		len_left -= (tag_len+2);
+	switch(chipset) {
+	case ESP_FIRMWARE_CHIP_ESP32:
+	case ESP_FIRMWARE_CHIP_ESP32C6:
+		adapter->chipset = chipset;
+		esp_info("Chipset=%s ID=%02x detected over SDIO\n", esp_chipname_from_id(chipset), chipset);
+		break;
+	case ESP_FIRMWARE_CHIP_ESP32S2:
+	case ESP_FIRMWARE_CHIP_ESP32S3:
+	case ESP_FIRMWARE_CHIP_ESP32C2:
+	case ESP_FIRMWARE_CHIP_ESP32C3:
+		esp_err("Chipset=%s ID=%02x not supported for SDIO\n", esp_chipname_from_id(chipset), chipset);
+		adapter->chipset = ESP_FIRMWARE_CHIP_UNRECOGNIZED;
+		break;
+	default:
+		esp_err("Unrecognized Chipset ID=%02x\n", chipset);
+		adapter->chipset = ESP_FIRMWARE_CHIP_UNRECOGNIZED;
+		break;
 	}
 
-	if (esp_add_card(adapter)) {
-		esp_err("network iterface init failed\n");
-		generate_slave_intr(context, BIT(ESP_CLOSE_DATA_PATH));
-	}
+	return ret;
+}
+
+int esp_adjust_spi_clock(struct esp_adapter *adapter, u8 spi_clk_mhz)
+{
+	/* SPI bus specific call, silently discard */
+	return 0;
 }
 
 void esp_deinit_interface_layer(void)
diff --git a/drivers/net/wireless/espressif/esp32-ng/esp_sdio_decl.h b/drivers/net/wireless/espressif/esp32-ng/esp_sdio_decl.h
index 1d2298917f0d5..df4caa3d15aac 100644
--- a/drivers/net/wireless/espressif/esp32-ng/esp_sdio_decl.h
+++ b/drivers/net/wireless/espressif/esp32-ng/esp_sdio_decl.h
@@ -65,10 +65,15 @@
 
 #define ESP_ADDRESS_MASK              0x3FF
 
-#define ESP_VENDOR_ID                 0x6666
-#define ESP_DEVICE_ID_1               0x2222
-#define ESP_DEVICE_ID_2               0x3333
-
+#if defined(CONFIG_TARGET_ESP32)
+	#define ESP_VENDOR_ID               0x6666
+	#define ESP_DEVICE_ID_1             0x2222
+	#define ESP_DEVICE_ID_2             0x3333
+#elif defined(CONFIG_TARGET_ESP32C6)
+	#define ESP_VENDOR_ID               0x0092
+	#define ESP_DEVICE_ID_1             0x6666
+	#define ESP_DEVICE_ID_2             0x7777
+#endif
 
 enum context_state {
 	ESP_CONTEXT_DISABLED = 0,
@@ -83,6 +88,7 @@ struct esp_sdio_context {
 	struct sk_buff_head    tx_q[MAX_PRIORITY_QUEUES];
 	u32                    rx_byte_count;
 	u32                    tx_buffer_count;
+	u32		       sdio_clk_mhz;
 };
 
 #endif
diff --git a/drivers/net/wireless/espressif/esp32-ng/esp_spi.c b/drivers/net/wireless/espressif/esp32-ng/esp_spi.c
index aca8b6757e111..c741d71981fc6 100644
--- a/drivers/net/wireless/espressif/esp32-ng/esp_spi.c
+++ b/drivers/net/wireless/espressif/esp32-ng/esp_spi.c
@@ -14,12 +14,14 @@
 #include "esp_bt_api.h"
 #include "esp_kernel_port.h"
 #include "esp_stats.h"
+#include "esp_utils.h"
+#include "esp_cfg80211.h"
 
 #define SPI_INITIAL_CLK_MHZ     10
-#define NUMBER_1M               1000000
 #define TX_MAX_PENDING_COUNT    100
 #define TX_RESUME_THRESHOLD     (TX_MAX_PENDING_COUNT/5)
 
+extern u32 raw_tp_mode;
 static struct sk_buff *read_packet(struct esp_adapter *adapter);
 static int write_packet(struct esp_adapter *adapter, struct sk_buff *skb);
 static void spi_exit(void);
@@ -31,7 +33,6 @@ volatile u8 host_sleep;
 static struct esp_spi_context spi_context;
 static char hardware_type = ESP_FIRMWARE_CHIP_UNRECOGNIZED;
 static atomic_t tx_pending;
-static uint8_t esp_reset_after_module_load;
 
 static struct esp_if_ops if_ops = {
 	.read		= read_packet,
@@ -156,143 +157,57 @@ static int write_packet(struct esp_adapter *adapter, struct sk_buff *skb)
 	return 0;
 }
 
-void process_event_esp_bootup(struct esp_adapter *adapter, u8 *evt_buf, u8 len)
+int esp_validate_chipset(struct esp_adapter *adapter, u8 chipset)
 {
-	/* Bootup event will be received whenever ESP is booted.
-	 * It is termed 'First bootup' when this event is received
-	 * the first time module loaded. It is termed 'Second & onward bootup' when
-	 * there is ESP reset (reason may be manual reset of ESP or any crash at ESP)
-	 */
-	u8 len_left = len, tag_len;
-	u8 *pos;
-	uint8_t iface_idx = 0;
-	uint8_t prio_q_idx = 0;
-
-	if (!adapter)
-		return;
-
-	if (!evt_buf)
-		return;
-
-	/* Second & onward bootup, cleanup and re-init the driver */
-	if (esp_reset_after_module_load)
-		set_bit(ESP_CLEANUP_IN_PROGRESS, &adapter->state_flags);
-
-	pos = evt_buf;
-
-	while (len_left) {
-
-		tag_len = *(pos + 1);
-
-		esp_info("EVENT: %d\n", *pos);
-
-		if (*pos == ESP_BOOTUP_CAPABILITY) {
-
-			adapter->capabilities = *(pos + 2);
-
-		} else if (*pos == ESP_BOOTUP_FW_DATA) {
-
-			if (tag_len != sizeof(struct fw_data))
-				esp_info("Length not matching to firmware data size\n");
-			else
-				if (process_fw_data((struct fw_data *)(pos + 2))) {
-					esp_remove_card(spi_context.adapter);
-					return;
-				}
-
-		} else if (*pos == ESP_BOOTUP_SPI_CLK_MHZ) {
-
-			adjust_spi_clock(*(pos + 2));
-			adapter->dev = &spi_context.esp_spi_dev->dev;
-
-		} else if (*pos == ESP_BOOTUP_FIRMWARE_CHIP_ID) {
-
-			hardware_type = *(pos+2);
-
-		} else if (*pos == ESP_BOOTUP_TEST_RAW_TP) {
-			process_test_capabilities(*(pos + 2));
-		} else {
-			esp_warn("Unsupported tag in event");
-		}
-
-		pos += (tag_len+2);
-		len_left -= (tag_len+2);
-	}
-
-	if ((hardware_type != ESP_FIRMWARE_CHIP_ESP32) &&
-	    (hardware_type != ESP_FIRMWARE_CHIP_ESP32S2) &&
-	    (hardware_type != ESP_FIRMWARE_CHIP_ESP32C3) &&
-	    (hardware_type != ESP_FIRMWARE_CHIP_ESP32S3)) {
-		esp_info("ESP chipset not recognized, ignoring [%d]\n", hardware_type);
-		hardware_type = ESP_FIRMWARE_CHIP_UNRECOGNIZED;
-	} else {
-		esp_info("ESP chipset detected [%s]\n",
-				hardware_type == ESP_FIRMWARE_CHIP_ESP32 ? "esp32" :
-				hardware_type == ESP_FIRMWARE_CHIP_ESP32S2 ? "esp32-s2" :
-				hardware_type == ESP_FIRMWARE_CHIP_ESP32C3 ? "esp32-c3" :
-				hardware_type == ESP_FIRMWARE_CHIP_ESP32S3 ? "esp32-s3" :
-				"unknown");
-	}
-
-	if (esp_reset_after_module_load) {
-
-		/* Second & onward bootup:
-		 *
-		 * SPI is software and not a hardware based module.
-		 * When bootup event is received, we should discard all prior commands,
-		 * old messages pending at network and re-initialize everything.
-		 *
-		 * Such handling is not required
-		 * 1. for SDIO
-		 *   as Removal of SDIO triggers complete Deinit and on SDIO insertion/
-		 *   detection, i.e., after probing, initialization is triggered
-		 *
-		 * 2. On first bootup (if counterpart of this else)
-		 *   First bootup event is received immediately after module insertion.
-		 *   As all network or cmds are init and clean for the first time,
-		 *   there is no need to re-init them
-		 */
-
-		for (prio_q_idx = 0; prio_q_idx < MAX_PRIORITY_QUEUES; prio_q_idx++) {
-			skb_queue_purge(&spi_context.tx_q[prio_q_idx]);
-		}
-
-		for (iface_idx = 0; iface_idx < ESP_MAX_INTERFACE; iface_idx++) {
-
-			struct esp_wifi_device *priv = adapter->priv[iface_idx];
-
-			if (!priv)
-				continue;
-
-			if (priv->scan_in_progress)
-				ESP_MARK_SCAN_DONE(priv, true);
+	int ret = 0;
 
-			if (priv->ndev &&
-			    wireless_dev_current_bss_exists(&priv->wdev)) {
-				CFG80211_DISCONNECTED(priv->ndev,
-						0, NULL, 0, false, GFP_KERNEL);
-			}
-		}
+	switch(chipset) {
+	case ESP_FIRMWARE_CHIP_ESP32:
+	case ESP_FIRMWARE_CHIP_ESP32S2:
+	case ESP_FIRMWARE_CHIP_ESP32S3:
+	case ESP_FIRMWARE_CHIP_ESP32C2:
+	case ESP_FIRMWARE_CHIP_ESP32C3:
+	case ESP_FIRMWARE_CHIP_ESP32C6:
+		adapter->chipset = chipset;
+		esp_info("Chipset=%s ID=%02x detected over SPI\n", esp_chipname_from_id(chipset), chipset);
+		break;
+	default:
+		esp_err("Unrecognized chipset ID=%02x\n", chipset);
+		adapter->chipset = ESP_FIRMWARE_CHIP_UNRECOGNIZED;
+		break;
+	}
+
+	return ret;
+}
 
-		esp_remove_card(adapter);
+int esp_deinit_module(struct esp_adapter *adapter)
+{
+	/* Second & onward bootup cleanup:
+	 *
+	 * SPI is software and not a hardware based module.
+	 * When bootup event is received, we should discard all prior commands,
+	 * old messages pending at network and re-initialize everything.
+	 */
+	uint8_t prio_q_idx, iface_idx;
 
-		for (prio_q_idx = 0; prio_q_idx < MAX_PRIORITY_QUEUES; prio_q_idx++) {
-			skb_queue_head_init(&spi_context.tx_q[prio_q_idx]);
-		}
+	for (prio_q_idx = 0; prio_q_idx < MAX_PRIORITY_QUEUES; prio_q_idx++) {
+		skb_queue_purge(&spi_context.tx_q[prio_q_idx]);
 	}
 
-	if (esp_add_card(adapter)) {
-		esp_err("network iterface init failed\n");
+	for (iface_idx = 0; iface_idx < ESP_MAX_INTERFACE; iface_idx++) {
+		struct esp_wifi_device *priv = adapter->priv[iface_idx];
+		esp_mark_scan_done_and_disconnect(priv, true);
 	}
 
-	process_capabilities(adapter);
-	print_capabilities(adapter->capabilities);
+	esp_remove_card(adapter);
 
+	for (prio_q_idx = 0; prio_q_idx < MAX_PRIORITY_QUEUES; prio_q_idx++) {
+		skb_queue_head_init(&spi_context.tx_q[prio_q_idx]);
+	}
 
-	esp_reset_after_module_load = 1;
+	return 0;
 }
 
-
 static int process_rx_buf(struct sk_buff *skb)
 {
 	struct esp_payload_header *header;
@@ -378,9 +293,11 @@ static void esp_spi_work(struct work_struct *work)
 				cb = (struct esp_skb_cb *)tx_skb->cb;
 				if (cb && cb->priv && atomic_read(&tx_pending) < TX_RESUME_THRESHOLD) {
 					esp_tx_resume(cb->priv);
-					#if TEST_RAW_TP
+#if TEST_RAW_TP
+					if (raw_tp_mode != 0) {
 						esp_raw_tp_queue_resume();
-					#endif
+					}
+#endif
 				}
 			}
 		}
@@ -499,7 +416,7 @@ static int spi_dev_init(int spi_clk_mhz)
 	struct spi_board_info esp_board = {{0}};
 	struct spi_master *master = NULL;
 
-	strlcpy(esp_board.modalias, "esp_spi", sizeof(esp_board.modalias));
+	strscpy(esp_board.modalias, "esp_spi", sizeof(esp_board.modalias));
 	esp_board.mode = SPI_MODE_2;
 	esp_board.max_speed_hz = spi_clk_mhz * NUMBER_1M;
 	esp_board.bus_num = 0;
@@ -675,7 +592,14 @@ static void adjust_spi_clock(u8 spi_clk_mhz)
 	}
 }
 
-int esp_init_interface_layer(struct esp_adapter *adapter)
+int esp_adjust_spi_clock(struct esp_adapter *adapter, u8 spi_clk_mhz)
+{
+	adjust_spi_clock(spi_clk_mhz);
+
+	return 0;
+}
+
+int esp_init_interface_layer(struct esp_adapter *adapter, u32 speed)
 {
 	if (!adapter)
 		return -EINVAL;
@@ -686,7 +610,10 @@ int esp_init_interface_layer(struct esp_adapter *adapter)
 	adapter->if_ops = &if_ops;
 	adapter->if_type = ESP_IF_TYPE_SPI;
 	spi_context.adapter = adapter;
-	spi_context.spi_clk_mhz = SPI_INITIAL_CLK_MHZ;
+	if (speed)
+		spi_context.spi_clk_mhz = speed;
+	else
+		spi_context.spi_clk_mhz = SPI_INITIAL_CLK_MHZ;
 
 	return spi_init();
 }
diff --git a/drivers/net/wireless/espressif/esp32-ng/esp_stats.c b/drivers/net/wireless/espressif/esp32-ng/esp_stats.c
index 4cd6a85ce97fc..231a913ff0ed7 100644
--- a/drivers/net/wireless/espressif/esp32-ng/esp_stats.c
+++ b/drivers/net/wireless/espressif/esp32-ng/esp_stats.c
@@ -179,21 +179,16 @@ void update_test_raw_tp_rx_stats(u16 len)
 }
 #endif
 
-void process_test_capabilities(u8 cap)
+void process_test_capabilities(u32 raw_tp_mode)
 {
 #if TEST_RAW_TP
-	esp_info("ESP peripheral test capabilities: 0x%x\n", cap);
-	if ((cap & ESP_TEST_RAW_TP) == ESP_TEST_RAW_TP) {
-		if ((cap & ESP_TEST_RAW_TP__ESP_TO_HOST) == ESP_TEST_RAW_TP__ESP_TO_HOST) {
-			start_test_raw_tp(ESP_TEST_RAW_TP__RX);
-			esp_info("start testing of ESP->Host raw throughput\n");
-		} else {
-			start_test_raw_tp(ESP_TEST_RAW_TP__TX);
-			esp_info("start testing of Host->ESP raw throughput\n");
-		}
-	} else {
-		esp_info("stop raw throuput test if running\n");
-		stop_test_raw_tp();
+	stop_test_raw_tp();
+	if (raw_tp_mode == ESP_TEST_RAW_TP_ESP_TO_HOST) {
+		start_test_raw_tp(ESP_TEST_RAW_TP__RX);
+		esp_info("start testing of ESP->Host raw throughput\n");
+	} else if (raw_tp_mode == ESP_TEST_RAW_TP_HOST_TO_ESP) {
+		start_test_raw_tp(ESP_TEST_RAW_TP__TX);
+		esp_info("start testing of Host->ESP raw throughput\n");
 	}
 	process_raw_tp_flags();
 #endif
diff --git a/drivers/net/wireless/espressif/esp32-ng/esp_stats.h b/drivers/net/wireless/espressif/esp32-ng/esp_stats.h
index 179266a0dc272..a508d2c6c7de2 100644
--- a/drivers/net/wireless/espressif/esp32-ng/esp_stats.h
+++ b/drivers/net/wireless/espressif/esp32-ng/esp_stats.h
@@ -10,11 +10,7 @@
 
 #include "esp.h"
 
-#ifdef CONFIG_TEST_RAW_TP
 #define TEST_RAW_TP 1
-#else
-#define TEST_RAW_TP 0
-#endif
 
 #if TEST_RAW_TP
 
diff --git a/drivers/net/wireless/espressif/esp32-ng/esp_wpa_utils.c b/drivers/net/wireless/espressif/esp32-ng/esp_utils.c
similarity index 55%
rename from drivers/net/wireless/espressif/esp32-ng/esp_wpa_utils.c
rename to drivers/net/wireless/espressif/esp32-ng/esp_utils.c
index 379bd1f7069be..56e2b64f36533 100644
--- a/drivers/net/wireless/espressif/esp32-ng/esp_wpa_utils.c
+++ b/drivers/net/wireless/espressif/esp32-ng/esp_utils.c
@@ -6,7 +6,7 @@
  *
  */
 #include "utils.h"
-#include "esp_wpa_utils.h"
+#include "esp_utils.h"
 
 
 int wpa_cipher_to_alg(int cipher)
@@ -31,33 +31,20 @@ int wpa_cipher_to_alg(int cipher)
 	return WIFI_WPA_ALG_NONE;
 }
 
-int wpa_cipher_key_len(int cipher)
+char * esp_chipname_from_id(int chipset_id)
 {
-	switch (cipher) {
-	case WLAN_CIPHER_SUITE_TKIP:
-#ifdef CONFIG_GCMP
-	case WLAN_CIPHER_GCMP_SUITE_256:
-#endif
-#ifdef CONFIG_GMAC
-	case WLAN_CIPHER_BIP_GMAC_SUITE_256:
-#endif
-		return 32;
-	case WLAN_CIPHER_SUITE_CCMP:
-#ifdef CONFIG_GCMP
-	case WLAN_CIPHER_SUITE_GCMP:
-#endif
-#ifdef CONFIG_GMAC
-	case WLAN_CIPHER_SUITE_BIP_GMAC_128:
-#endif
-#if 0
-	case WLAN_CIPHER_SUITE_AES_128_CMAC:
-		return 16;
-#endif
-	case WLAN_CIPHER_SUITE_WEP104:
-		return 13;
-	case WLAN_CIPHER_SUITE_WEP40:
-		return 5;
-	}
+	if (chipset_id == ESP_FIRMWARE_CHIP_ESP32)
+		return "ESP32";
+	if (chipset_id == ESP_FIRMWARE_CHIP_ESP32S2)
+		return "ESP32-S2";
+	if (chipset_id == ESP_FIRMWARE_CHIP_ESP32S3)
+		return "ESP32-S3";
+	if (chipset_id == ESP_FIRMWARE_CHIP_ESP32C2)
+		return "ESP32-C2";
+	if (chipset_id == ESP_FIRMWARE_CHIP_ESP32C3)
+		return "ESP32-C3";
+	if (chipset_id == ESP_FIRMWARE_CHIP_ESP32C6)
+		return "ESP32-C6";
 
-	return 0;
+	return "Unknown Chip";
 }
diff --git a/drivers/net/wireless/espressif/esp32-ng/esp_wpa_utils.h b/drivers/net/wireless/espressif/esp32-ng/esp_utils.h
similarity index 92%
rename from drivers/net/wireless/espressif/esp32-ng/esp_wpa_utils.h
rename to drivers/net/wireless/espressif/esp32-ng/esp_utils.h
index 3d7d0f810c528..38b3a651bc1d3 100644
--- a/drivers/net/wireless/espressif/esp32-ng/esp_wpa_utils.h
+++ b/drivers/net/wireless/espressif/esp32-ng/esp_utils.h
@@ -5,8 +5,8 @@
  * SPDX-FileCopyrightText: 2015-2023 Espressif Systems (Shanghai) CO LTD
  *
  */
-#ifndef __esp_wpa_utils_h_
-#define __esp_wpa_utils_h_
+#ifndef __esp_utils_h_
+#define __esp_utils_h_
 
 #include "esp.h"
 
@@ -37,5 +37,6 @@ enum wpa_alg {
 
 int wpa_cipher_to_alg(int cipher);
 
+char * esp_chipname_from_id(int chipset_id);
 
 #endif
diff --git a/drivers/net/wireless/espressif/esp32-ng/main.c b/drivers/net/wireless/espressif/esp32-ng/main.c
index 263dafcf2abcf..b30e81a3ec17f 100644
--- a/drivers/net/wireless/espressif/esp32-ng/main.c
+++ b/drivers/net/wireless/espressif/esp32-ng/main.c
@@ -22,14 +22,23 @@
 #include "esp_cfg80211.h"
 #include "esp_stats.h"
 
+#define RELEASE_VERSION "1.0.3"
 #define HOST_GPIO_PIN_INVALID -1
 static int resetpin = HOST_GPIO_PIN_INVALID;
+static u32 clockspeed = 0;
 extern u8 ap_bssid[MAC_ADDR_LEN];
 extern volatile u8 host_sleep;
+u32 raw_tp_mode = 0;
 
 module_param(resetpin, int, S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH);
 MODULE_PARM_DESC(resetpin, "Host's GPIO pin number which is connected to ESP32's EN to reset ESP32 device");
 
+module_param(clockspeed, uint, S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH);
+MODULE_PARM_DESC(clockspeed, "Hosts clock speed in MHz");
+
+module_param(raw_tp_mode, uint, S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH);
+MODULE_PARM_DESC(raw_tp_mode, "Mode choosed to test raw throughput");
+
 static void deinit_adapter(void);
 
 
@@ -203,12 +212,8 @@ void print_capabilities(u32 cap)
 	}
 }
 
-void process_capabilities(struct esp_adapter *adapter)
+void init_bt(struct esp_adapter *adapter)
 {
-	esp_info("ESP peripheral capabilities: 0x%x\n", adapter->capabilities);
-
-	/* Reset BT */
-	esp_deinit_bt(adapter);
 
 	if ((adapter->capabilities & ESP_BT_SPI_SUPPORT) ||
 		(adapter->capabilities & ESP_BT_SDIO_SUPPORT)) {
@@ -220,7 +225,7 @@ void process_capabilities(struct esp_adapter *adapter)
 
 static int check_esp_version(struct fw_version *ver)
 {
-	esp_info("esp32: ESP Firmware version: %u.%u.%u\n",
+	esp_info("ESP Firmware version: %u.%u.%u\n",
 			ver->major1, ver->major2, ver->minor);
 	if (!ver->major1) {
 		esp_err("Incompatible ESP firmware release detected, Please use correct ESP-Hosted branch/compatible release\n");
@@ -252,18 +257,89 @@ static void print_reset_reason(uint32_t reason)
 	}
 }
 
-int process_fw_data(struct fw_data *fw_p)
+static int process_fw_data(struct fw_data *fw_p, int tag_len)
 {
-	if (!fw_p) {
-		esp_err("Incomplete/incorrect bootup event received\n");
+	if (tag_len != sizeof(struct fw_data)) {
+		esp_err("Length not matching to firmware data size\n");
 		return -1;
 	}
 
 	esp_info("ESP chipset's last reset cause:\n");
 	print_reset_reason(le32_to_cpu(fw_p->last_reset_reason));
+
 	return check_esp_version(&fw_p->version);
 }
 
+int process_event_esp_bootup(struct esp_adapter *adapter, u8 *evt_buf, u8 len)
+{
+	int len_left = len, tag_len, ret = 0;
+	u8 *pos;
+
+	if (!adapter || !evt_buf)
+		return -1;
+
+	if (len_left >= 64) {
+		esp_info("ESP init event len looks unexpected: %u (>=64)\n", len_left);
+		esp_info("You probably facing timing mismatch at transport layer\n");
+	}
+
+	clear_bit(ESP_INIT_DONE, &adapter->state_flags);
+	/* Deinit module if already initialized */
+	esp_deinit_module(adapter);
+
+	pos = evt_buf;
+
+	while (len_left > 0) {
+		tag_len = *(pos + 1);
+
+		esp_info("Bootup Event tag: %d\n", *pos);
+
+		switch (*pos) {
+		case ESP_BOOTUP_CAPABILITY:
+			adapter->capabilities = *(pos + 2);
+			break;
+		case ESP_BOOTUP_FIRMWARE_CHIP_ID:
+			ret = esp_validate_chipset(adapter, *(pos + 2));
+			break;
+		case ESP_BOOTUP_FW_DATA:
+			ret = process_fw_data((struct fw_data *)(pos + 2), tag_len);
+			break;
+		case ESP_BOOTUP_SPI_CLK_MHZ:
+			ret = esp_adjust_spi_clock(adapter, *(pos + 2));
+			break;
+		default:
+			esp_warn("Unsupported tag=%x in bootup event\n", *pos);
+		}
+
+		if (ret < 0) {
+			esp_err("failed to process tag=%x in bootup event\n", *pos);
+			return -1;
+		}
+		pos += (tag_len + 2);
+		len_left -= (tag_len + 2);
+	}
+
+	if (esp_add_card(adapter)) {
+		esp_err("network iterface init failed\n");
+		return -1;
+	}
+	init_bt(adapter);
+
+	if (raw_tp_mode !=0) {
+#if TEST_RAW_TP
+		process_test_capabilities(raw_tp_mode);
+		esp_init_raw_tp(adapter);
+#else
+		esp_err("RAW TP mode selected but not enabled\n");
+		return -1;
+#endif
+	}
+	set_bit(ESP_INIT_DONE, &adapter->state_flags);
+	print_capabilities(adapter->capabilities);
+
+	return 0;
+}
+
 static int esp_open(struct net_device *ndev)
 {
 	return 0;
@@ -271,9 +347,6 @@ static int esp_open(struct net_device *ndev)
 
 static int esp_stop(struct net_device *ndev)
 {
-	struct esp_wifi_device *priv = netdev_priv(ndev);
-
-	ESP_MARK_SCAN_DONE(priv, true);
 	return 0;
 }
 
@@ -392,10 +465,8 @@ void esp_init_priv(struct net_device *ndev)
 			INTERFACE_HEADER_PADDING, 4);
 }
 
-static int add_network_iface(void)
+static int esp_add_network_ifaces(struct esp_adapter *adapter)
 {
-	int ret = 0;
-	struct esp_adapter *adapter = esp_get_adapter();
 	struct wireless_dev *wdev = NULL;
 
 	if (!adapter) {
@@ -403,12 +474,6 @@ static int add_network_iface(void)
 		return -EINVAL;
 	}
 
-	ret = esp_cfg80211_register(adapter);
-	if (ret) {
-		esp_err("Failed to register with cfg80211 (err code 0x%x)\n", ret);
-		return ret;
-	}
-
 	rtnl_lock();
 	wdev = esp_cfg80211_add_iface(adapter->wiphy, "espsta%d", 1, NL80211_IFTYPE_STATION, NULL);
 	rtnl_unlock();
@@ -420,16 +485,22 @@ static int add_network_iface(void)
 	return -1;
 }
 
+int esp_init_raw_tp(struct esp_adapter *adapter)
+{
+	RET_ON_FAIL(cmd_init_raw_tp_task_timer(adapter->priv[ESP_STA_NW_IF]));
+	return 0;
+}
+
 int esp_add_card(struct esp_adapter *adapter)
 {
 	RET_ON_FAIL(esp_commands_setup(adapter));
-
-	RET_ON_FAIL(add_network_iface());
+	RET_ON_FAIL(esp_add_wiphy(adapter));
+	RET_ON_FAIL(esp_add_network_ifaces(adapter));
 
 	return 0;
 }
 
-void esp_remove_network_interfaces(struct esp_adapter *adapter)
+static int esp_remove_network_ifaces(struct esp_adapter *adapter)
 {
 	uint8_t iface_idx = 0;
 	struct net_device *ndev = NULL;
@@ -438,60 +509,89 @@ void esp_remove_network_interfaces(struct esp_adapter *adapter)
 	for (iface_idx = 0; iface_idx < ESP_MAX_INTERFACE; iface_idx++) {
 
 		priv = adapter->priv[iface_idx];
-
 		if (!priv)
 			continue;
-
 		if (!test_bit(ESP_NETWORK_UP, &priv->priv_flags))
 			continue;
 
-		/* stop and unregister network */
 		ndev = priv->ndev;
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 12, 0)
+		if (ndev)
+			ndev->needs_free_netdev = true;
+		rtnl_lock();
+		wiphy_lock(adapter->wiphy);
+		cfg80211_unregister_wdev(&priv->wdev);
+		wiphy_unlock(adapter->wiphy);
+		rtnl_unlock();
+#else
+		if (ndev && ndev->reg_state == NETREG_REGISTERED) {
+			unregister_netdev(ndev);
+			free_netdev(ndev);
+			ndev = NULL;
+		}
+#endif
+		adapter->priv[iface_idx] = NULL;
+	}
 
-		if (ndev) {
+	return 0;
+}
 
-			if (netif_carrier_ok(ndev))
-				netif_carrier_off(ndev);
+static int stop_network_iface(struct esp_wifi_device *priv)
+{
+	struct net_device *ndev;
 
-			netif_device_detach(ndev);
+	if (!priv)
+		return 0;
 
-			if (ndev->reg_state == NETREG_REGISTERED) {
-				unregister_inetaddr_notifier(&(adapter->priv[0]->nb));
-				unregister_netdev(ndev);
-				free_netdev(ndev);
-				ndev = NULL;
-			}
-		}
-		clear_bit(ESP_NETWORK_UP, &priv->priv_flags);
-	}
+	if (!test_bit(ESP_NETWORK_UP, &priv->priv_flags))
+		return 0;
+
+	esp_mark_scan_done_and_disconnect(priv, false);
+	esp_port_close(priv);
+
+	/* stop and unregister network */
+	ndev = priv->ndev;
 
-	if (adapter->wiphy) {
+	if (ndev) {
+		netif_carrier_off(ndev);
+		netif_device_detach(ndev);
 
-		wiphy_unregister(adapter->wiphy);
-		wiphy_free(adapter->wiphy);
-		adapter->wiphy = NULL;
+		unregister_inetaddr_notifier(&(priv->nb));
 	}
+
+	return 0;
 }
 
-int esp_remove_card(struct esp_adapter *adapter)
+int esp_stop_network_ifaces(struct esp_adapter *adapter)
 {
 	uint8_t iface_idx = 0;
 
-	if (!adapter) {
-		return 0;
+	for (iface_idx = 0; iface_idx < ESP_MAX_INTERFACE; iface_idx++) {
+		stop_network_iface(adapter->priv[iface_idx]);
 	}
 
-	esp_deinit_bt(adapter);
+	rtnl_lock();
+	if (adapter->wiphy)
+		cfg80211_shutdown_all_interfaces(adapter->wiphy);
 
-	esp_commands_teardown(adapter);
+	rtnl_unlock();
 
-	esp_remove_network_interfaces(adapter);
+	return 0;
+}
 
-	for (iface_idx = 0; iface_idx < ESP_MAX_INTERFACE; iface_idx++) {
-		esp_port_close(adapter->priv[iface_idx]);
-		adapter->priv[iface_idx] = NULL;
+int esp_remove_card(struct esp_adapter *adapter)
+{
+	if (!adapter) {
+		return 0;
 	}
 
+	esp_stop_network_ifaces(adapter);
+	/* BT may have been initialized after fw bootup event, deinit it */
+	esp_deinit_bt(adapter);
+	esp_commands_teardown(adapter);
+	esp_remove_network_ifaces(adapter);
+	esp_remove_wiphy(adapter);
+
 	return 0;
 }
 
@@ -531,7 +631,7 @@ static void process_esp_bootup_event(struct esp_adapter *adapter,
 		return;
 	}
 
-	esp_info("\nReceived ESP bootup event\n");
+	esp_info("Received ESP bootup event\n");
 	process_event_esp_bootup(adapter, evt->data, evt->len);
 }
 
@@ -541,7 +641,7 @@ static int process_internal_event(struct esp_adapter *adapter,
 	struct event_header *header = NULL;
 
 	if (!skb || !adapter) {
-		esp_err("esp32: Incorrect event data!\n");
+		esp_err("Incorrect event data!\n");
 		return -1;
 	}
 
@@ -683,9 +783,11 @@ static void process_rx_packet(struct esp_adapter *adapter, struct sk_buff *skb)
 			dev_kfree_skb_any(skb);
 
 	} else if (payload_header->if_type == ESP_TEST_IF) {
-		#if TEST_RAW_TP
+#if TEST_RAW_TP
+		if (raw_tp_mode != 0) {
 			update_test_raw_tp_rx_stats(len);
-		#endif
+		}
+#endif
 		dev_kfree_skb_any(skb);
 	} else {
 		dev_kfree_skb_any(skb);
@@ -884,7 +986,7 @@ static int __init esp_init(void)
 		return -EFAULT;
 
 	/* Init transport layer */
-	ret = esp_init_interface_layer(adapter);
+	ret = esp_init_interface_layer(adapter, clockspeed);
 
 	if (ret != 0) {
 		deinit_adapter();
@@ -897,7 +999,9 @@ static void __exit esp_exit(void)
 {
 	uint8_t iface_idx = 0;
 #if TEST_RAW_TP
-	test_raw_tp_cleanup();
+	if (raw_tp_mode != 0) {
+		test_raw_tp_cleanup();
+	}
 #endif
 	for (iface_idx = 0; iface_idx < ESP_MAX_INTERFACE; iface_idx++) {
 		cmd_deinit_interface(adapter.priv[iface_idx]);
@@ -916,6 +1020,6 @@ MODULE_AUTHOR("Amey Inamdar <amey.inamdar@espressif.com>");
 MODULE_AUTHOR("Mangesh Malusare <mangesh.malusare@espressif.com>");
 MODULE_AUTHOR("Yogesh Mantri <yogesh.mantri@espressif.com>");
 MODULE_DESCRIPTION("Wifi driver for ESP-Hosted solution");
-MODULE_VERSION("0.1");
+MODULE_VERSION(RELEASE_VERSION);
 module_init(esp_init);
 module_exit(esp_exit);
diff --git a/drivers/net/wireless/espressif/esp32-ng/utils.h b/drivers/net/wireless/espressif/esp32-ng/utils.h
index b588ca673b914..02082e63a2529 100644
--- a/drivers/net/wireless/espressif/esp32-ng/utils.h
+++ b/drivers/net/wireless/espressif/esp32-ng/utils.h
@@ -6,6 +6,10 @@
 
 #define pr_fmt(fmt) "%s:%s: " fmt, KBUILD_MODNAME, __func__
 
+#ifndef NUMBER_1M
+#define NUMBER_1M 1000000
+#endif
+
 #ifndef MAC2STR
 #define MAC2STR(a) (a)[0], (a)[1], (a)[2], (a)[3], (a)[4], (a)[5]
 #define MACSTR "%02x:%02x:%02x:%02x:%02x:%02x"
-- 
2.55.0

